### PR TITLE
Make groovy parsing of console log configurable

### PR DESCRIPTION
**Pull request by pjdarton**
[JENKINS-54832](https://issues.jenkins.io/browse/JENKINS-54832) alludes to security concerns regarding allowing (admin-configured) groovy parser code to run on the Jenkins master node, and that this is why the groovy parsers are unable to scan the build console log.
My (limited) understanding is that this is because the parser code is not (currently) run within a security-sandbox and therefore there are risks involved in allowing such code to run on the master (where the console log would be scanned).
However, a Jenkins admin _may_ decide that, as they control the parser code and can decide what code to allow in or not (just as they decide what groovy methods to permit in a sandboxed environment), they wish to allow this code to run on the master unhindered.

This PR allows such a choice to be a tick-box configurable option within the "Groovy Based Warnings Parsers" section of the Jenkins global configuration (Manage Jenkins -> Configure System):
image
It defaults to false (un-ticked) in order to ensure that existing users who upgrade are unaffected.

With the option un-ticked, groovy parsers missing a pattern will (now) indicate that a value is required:
image
Previously the first indication was when the build threw an exception and logged `FATAL: Static analysis tool .... cannot scan console log output, please define a file pattern`

With the option ticked, groovy parsers are allowed to not specify a pattern and the web form indicates this:
image
...and any build using the parser will have its console logged as expected instead of throwing an exception.

TL;DR: This lets the Jenkins administrator decide.

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java [hunk 1]
```diff
@@ -294,6 +294,9 @@ public FormValidation doCheckPattern(@AncestorInPath final AbstractProject<?, ?>
             if (!JENKINS.hasPermission(Item.CONFIGURE, project)) {
                 return FormValidation.ok();
             }
+            if ((pattern==null || pattern.trim().isEmpty()) && !canScanConsoleLog()) {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:297]:
```suggestion
            if ((pattern==null || pattern.trim().isEmpty()) && !canScanConsoleLog()) {
```

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java [hunk 2]
```diff
@@ -294,6 +294,9 @@ public FormValidation doCheckPattern(@AncestorInPath final AbstractProject<?, ?>
             if (!JENKINS.hasPermission(Item.CONFIGURE, project)) {
                 return FormValidation.ok();
             }
+            if ((pattern == null || pattern.trim().isEmpty()) && !canScanConsoleLog()) {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:297]: This will break existing parsers that provide a default pattern like CheckStyle.

It would make sense to provide that improved check in a method like `hasPattern` (or similar).

File: plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html [hunk 3]
```diff
@@ -2,4 +2,7 @@
     A pattern is defined by an <a href="https://ant.apache.org/manual/Types/fileset.html">Ant Fileset ''includes''</a>
     setting that specifies the report files to scan for issues. Multiple patterns can be separated by space or comma.
     Note that such a pattern is resolved in Jenkins' workspace, so the paths must be relative only.
- </div>
+    <p>
+    If no pattern is specified then the build's console text will be scanned instead.
+    </p>
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html:7]: This is not correct see comment above about CheckStyle

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/config.jelly [hunk 4]
```diff
@@ -5,10 +5,20 @@
...
   </f:entry>
 
-  <f:entry title="${%title.pattern}" field="pattern"
-           description="${%description.pattern('http://ant.apache.org/manual/Types/fileset.html')}">
-    <c:safe-textbox/>
-  </f:entry>
+  <j:choose>
+    <j:when test="${descriptor.canScanConsoleLog()}">
+      <f:entry title="${%title.pattern}" field="pattern"
+               description="${%description.consoleOrPattern}">
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/config.jelly:11]:
```suggestion
               description="${%description.consoleOrPattern('http://ant.apache.org/manual/Types/fileset.html')}">
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/config.jelly [hunk 5]
```diff
@@ -5,10 +5,20 @@
...
+        <c:safe-textbox/>
+      </f:entry>
+    </j:when>
+    <j:otherwise>
+      <f:entry title="${%title.pattern}" field="pattern"
+               description="${%description.pattern}">
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/config.jelly:17]:
```suggestion
               description="${%description.pattern('http://ant.apache.org/manual/Types/fileset.html')}">
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html [hunk 6]
```diff
@@ -1 +0,0 @@
-Note that due to security reasons scanning the console log using a Groovy script is not allowed.
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html:1]: Rather than deleting this file it would make sense to copy the help from the other `help-pattern.html` file.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly [hunk 7]
```diff
@@ -7,10 +7,13 @@
         <f:entry title="${%Parsers}" description="${%title.description}">
           <i:repeatable field="parsers"/>
         </f:entry>
+        <f:entry field="consoleLogScanningPermitted" title="${%Permit Parsing of Console?}" description="${%consoleLogScanningPermitted.description}">
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly:10]: Can you please use a property here as well for the `title`, I'm not sure how the question mark works in properties files.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.properties [hunk 8]
```diff
@@ -1 +1,2 @@
 title.description=Create new parsers for the warnings plug-in based on a regular expression and a Groovy script.
+consoleLogScanningPermitted.description=Allow these parsers to act upon the build console.
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.properties:2]: Jenkins's security team mentioned that it would be helpful to make the text from your help file visible here as well (`help-canScanConsoleLog.html`). At least that checking the box can introduce a (potential) security risk.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<div>
+  <p>
+    Console Log parsing has to be processed directly on the Jenkins master
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html:3]:
```suggestion
    Console Log parsing has to be processed directly on the Jenkins controller
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    which means running the parser groovy code (configured here) directly on the master which <em>CAN BE UNSAFE</em> from a security point of view.
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html:4]:
```suggestion
    which means running the parser groovy code (configured here) directly on the controller which <em>CAN BE UNSAFE</em> from a security point of view.
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  </p>
+    e.g. if your groovy code is poorly written.
+  <p>
+    If you do not accept this risk then you should not enable this feature
+    ...
+    but if you do not enable this feature then the groovy parsers that you define will only be able to parse files in the build workspace
+    (safely processed on the nodes responsible for the builds rather than natively on the master)
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-canScanConsoleLog.html:11]:
```suggestion
    (safely processed on the agents responsible for the builds rather than natively on the controller)
```

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks for adding that enhancement! 

Can you please add a small test that shows that scanning the console log on the controller really works (a copy of `shouldRemoveConsoleLogNotesBeforeRemovingColorCodes` for instance with the Groovy Pep8 parser)?

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java (hunk 2 above)
Comment (pjdarton) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:297]: Apologies - I'm being a bit myopic and focussed purely on the Groovy use case...
How about:
```
            final boolean thereIsNoPatternConfigured = pattern == null || pattern.trim().isEmpty();
            final boolean thereIsNoDefaultPattern = getPattern().trim().isEmpty();
            if (thereIsNoPatternConfigured && thereIsNoDefaultPattern && !canScanConsoleLog()) {
                return FormValidation.error(Messages.ReportScanningTool_PatternIsEmptyAndConsoleParsingDisabled());
            }
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html (hunk 3 above)
Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html:7]: Hmm, yes...
I think that it would be beneficial to tell people this additional information _when it is true_ though.
I'll see if I can figure out a way of making it conditional on `getPattern()` returning an empty string - I think it's possible to have the help generated by groovy instead of simple HTML...

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html (hunk 6 above)
Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html:1]: Ideally, I'd like the other file to work for both cases 😉 
However, if I can't make it do that then, yes, I'll make a GroovyScript-specific version here.

Comment (uhafner) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html:1]: I tried that but it seems that Jenkins core does not support inheritance on those files. So I ended up copying the same files to all the folders.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java (hunk 2 above)
Comment (uhafner) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:297]: I think you need to write a method (or variable):
```
boolean hasPattern() {
    return StringUtils.isNotEmpty(getPattern()) || StringUtils.isNotEmpty(getDescriptor().getPattern());
}
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.properties (hunk 8 above)
Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.properties:2]: I'd prefer not to clutter the config page with a duplication of the help info.
...but I can do better than that - I'll add a doCheck method that'll show a warning when the checkbox is ticked that tells people to go read the help.
I suspect that a yellow warning will get more attention than static text.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html (hunk 6 above)
Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/GroovyScript/help-pattern.html:1]: I've experimented and I've found that, with the GroovyScript's help-pattern.html gone, Jenkins does inherit.
So I've made the inherited one flexible and able to cope with all 3 scenarios (default pattern, optional pattern, mandatory pattern) and I've checked that this works on e.g. both GroovyScript, CCM and others.

I did find that if you define a config.jelly file then that'll be used _instead_ of the parent's config.jelly but it's able to reference help files in the parent without a problem.

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java (hunk 2 above)
Comment (pjdarton) [reply plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java]: Done - I added a "hasDefaultPattern()" method and used that.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html (hunk 3 above)
Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/help-pattern.html]: I experimented and found that a help-pattern.jelly file isn't passed enough information to dynamically decide what to do.
So instead I extended the existing choose/when/otherwise pattern in the config.jelly to switch between 3 different help files as well as descriptions.
I've checked and this works for e.g. CCM, Brakeman etc as well as GroovyScript - the help, description and validation now document/enforce the same logic.

**Comment by pjdarton** [comment]
I believe I've addressed all the issues _except_ adding a "small test that shows that scanning the console log".
I'm not sure how to do that as, in my IDE, `StepsITest` is riddled with compilation errors because it can't `import static io.jenkins.plugins.analysis.core.assertions.Assertions.*` and, despite some searching, I can't find where that package/class is defined either.
...and if I do `mvn install` (or even `mvn '-Dtest=StepsITest#shouldShowWarningsOfGroovyParser' test`) on the command-line then it fails for other reasons (after a lengthy delay, probably because it doesn't like running on my Linux VM whose working directory is a Windows filesystem).
i.e. My dev environment isn't up to this kind of task and I'm not sure how to fix it.

What I'd like to do w.r.t. a unit-test is something like...
- Rename `shouldShowWarningsOfGroovyParser` to `shouldShowWarningsOfGroovyParserWhenScanningFileInWorkspace`
- Create a new test `shouldShowWarningsOfGroovyParserWhenScanningConsoleLogWhenThatIsPermitted`.
Read in `"pep8Test.txt"` as a series of lines, turn each into an `echo` statement to go into the `asStage` data before the `scanForIssues` part.
Have the scan specify `pattern:''` so it'll only look at the console.
Call `configuration.setConsoleLogScanningPermitted(true)` so this is permitted.
Run the build and expect success.
Expect to get the same scan results as for the original `shouldShowWarningsOfGroovyParser` test.
- Create another test called `shouldFailUsingGroovyParserToScanConsoleLogWhenThatIsForbidden` that's nearly identical to the previous one but which does NOT call `configuration.setConsoleLogScanningPermitted(true)` and which expects the build to fail (with no scanned results).

However, because I've got issues with my development environment, I'd be mostly "coding blind" with a 45-minute turnaround time between "commit & push" and seeing the results which just isn't viable; I think I'm going to struggle to get any non-trivial unit-test working without help 😞

**Comment by uhafner** [comment]
Which ID are you using? In IntelliJ you sometimes need to manually add the folder `target/generated-test-sources/assertj-assertions` (see https://github.com/uhafner/warnings-ng-plugin-devenv#installation-steps) as test source folders.

**Comment by pjdarton** [comment]
Thanks for that hint - that sorted it and got me going again 😁 

FYI there's no (obvious) link to the "Warnings Plugin - Development Environment" from the main README.md file.
I've (now) spotted that there's a link from the contributions page ... but I didn't read that page before 😞 
Might be an idea to link to the Dev-env and/or contributions page from the main README.md file, e.g. in the "If your tool is not yet supported" paragraph so the folks have more of an opportunity to spot it (or to feel more foolish about not having seen it 😉)

File: plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java [hunk 12]
```diff
@@ -84,6 +87,50 @@ public void setParsers(final List<GroovyParser> parsers) {
...
+     * {@link #setConsoleLogScanningPermitted(boolean)}.
+     * 
+     * @param value The current value.
+     * @return {@link FormValidation#ok()} if all is well, else a warning or error.
+     */
+    // Maintenance note: This is NOT annotated with @POST because it does not leak secrets.
+    // Similarly, there's no need to permission-check either.
+    // If it is ever changed to do anything non-trivial that might leak secrets then
+    // it must be changed to @POST and checkMethod="post" be added to the checkbox.
+    // There is a matching entry in archunit_ignore_patterns.txt to permit this.
```
Comment (pjdarton) [inline plugin/src/main/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration.java:126]: Note: I did try adding checkMethod="post" to the checkbox element, but it didn't work - it looks like checkbox does not support that attribute (and there's no safe-checkbox either).
i.e. right now, we have no other option but to use a non-POST callback ... but that's OK because this doCheck method does not leak data (you either get a fxied-text warning, or you get no error - there's no side-effects).

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly [hunk 13]
```diff
@@ -7,10 +7,13 @@
...
           <i:repeatable field="parsers"/>
         </f:entry>
+        <f:entry field="consoleLogScanningPermitted" title="${%consoleLogScanningPermitted.title}" description="${%consoleLogScanningPermitted.description}">
+          <f:checkbox/>
+        </f:entry>
       </j:when>
       <j:otherwise>
         <f:block>
-          You have no permission to edit this section (required permission: RUN_SCRIPTS).
+          You have no permission to edit this section (required permission: ADMINISTER).
```
Comment (pjdarton) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly:16]: FYI the underlying code checks for Jenkins.ADMINISTER - there's no check for RUN_SCRIPTS.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    which means running the parser groovy code (configured here) directly on the controller which <em>CAN BE UNSAFE</em> from a security point of view.
+  </p>
+    e.g. if your groovy code is poorly written.
+  <p>
+    If you do not accept this risk then you should not enable this feature
+    ...
+    but if you do not enable this feature then the groovy parsers that you define will only be able to parse files in the build workspace
+    (safely processed on the agents responsible for the builds rather than natively on the controller)
+    rather than the build console log itself.
+  </p>
```
Comment (pjdarton) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:13]: This text should be reviewed by someone who's 100% aware of the Jenkins-security-team's concerns about this.

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 15]
```diff
@@ -19,14 +20,14 @@
     private static final List<GroovyParser> PARSERS = Collections.singletonList(mock(GroovyParser.class));
 
     @Test
-    void shouldHaveNoRootFoldersWhenCreated() {
+    void shouldHaveNoParsersWhenCreated() {
```
Comment (pjdarton) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:23]: I've renamed this test case to better describe what it's testing.
(I suspect there was some copy/paste of test names when this class was originally created)

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java [hunk 16]
```diff
@@ -19,14 +20,14 @@
...
         ParserConfiguration configuration = new ParserConfiguration(mock(GlobalConfigurationFacade.class));
 
         assertThat(configuration.getParsers()).isEmpty();
     }
 
     @Test
-    void shouldSaveConfigurationIfFoldersAreAdded() {
+    void shouldSaveConfigurationIfParsersAreAdded() {
```
Comment (pjdarton) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/groovy/ParserConfigurationTest.java:30]: I've renamed this test case to better describe what it's testing.
(I suspect there was some copy/paste of test names when this class was originally created)

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java [hunk 17]
```diff
@@ -732,6 +734,63 @@ public void shouldShowWarningsOfGroovyParser() {
...
+        WorkflowJob job = createPipeline();
+        ArrayList<String> stages = new ArrayList<>();
+        catFileContentsByAddingEchosSteps(stages, "pep8Test.txt");
+        stages.add("def groovy = scanForIssues "
+                + "tool: groovyScript(parserId: 'groovy-pep8', pattern:'', reportEncoding:'UTF-8')");
+        stages.add("publishIssues issues:[groovy]");
+        job.setDefinition(asStage(stages.toArray(new String[0])));
+
+        ParserConfiguration configuration = ParserConfiguration.getInstance();
+        configuration.setConsoleLogScanningPermitted(true);
```
Comment (pjdarton) [inline plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:756]: CPD is complaining that it `Found duplicated code` and I don't intend to fix that - many unit tests are fairly repetitive by nature.
I've extracted some code out into common methods to reduce duplication but there will inevitably be some commonality between related tests.

File: plugin/src/test/resources/archunit_ignore_patterns.txt [hunk 18]
```diff
@@ -1 +1,6 @@
-// empty up to now, see https://www.archunit.org/userguide/html/000_Index.html#_ignoring_violations
\ No newline at end of file
+# see https://www.archunit.org/userguide/html/000_Index.html#_ignoring_violations
+
+# ParserConfiguration.doCheckConsoleLogScanningPermitted does not require @POST or a permission check.
+# As-implemented, it does not leak data, even if called illegally.
+.*ParserConfiguration\.doCheckConsoleLogScanningPermitted.*POST.*
+.*Permission.*ParserConfiguration\.doCheckConsoleLogScanningPermitted.*
```
Comment (pjdarton) [inline plugin/src/test/resources/archunit_ignore_patterns.txt:6]: I tried to avoid having to add any ignores, but there's no reason why that specific doCheck method requires a POST - that's only needed when a doCheck method can leak information.

**Comment by pjdarton** [comment]
OK, I think this is all sorted now; I belive I've addressed all the previous review comments and sorted all the CPD/CheckStyle/PMD/coverage issues too.

**Comment by uhafner** [comment]
Thanks! I'll hopefully find time for a review this evening.

File: plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java (hunk 17 above)
Comment (uhafner) [reply plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:756]: Yes, this is ok! There is no quality gate enforced for CPD anyway, it is just to provide some feedback.

**Comment by uhafner** [comment]
> FYI there's no (obvious) link to the "Warnings Plugin - Development Environment" from the main README.md file.
I've (now) spotted that there's a link from the contributions page ... but I didn't read that page before 😞
Might be an idea to link to the Dev-env and/or contributions page from the main README.md file, e.g. in the "If your tool is not yet supported" paragraph so the folks have more of an opportunity to spot it (or to feel more foolish about not having seen it 😉

Yes, I need to add a section to the README. Thanks for pointing that out.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html [hunk 19]
```diff
@@ -0,0 +1,14 @@
+<div>
+  <p>
+    Console Log parsing has to be processed directly on the Jenkins controller
+    which means running the parser groovy code (configured here) directly on the controller which <em>CAN BE UNSAFE</em> from a security point of view.
+  </p>
+    e.g. if your groovy code is poorly written.
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:6]: Is this sentence intended to be in the middle of 2 `<p>` tags?

Bildschirmfoto 2021-06-29 um 11 11 28

File: plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java [hunk 20]
```diff
@@ -294,10 +294,29 @@ public FormValidation doCheckPattern(@AncestorInPath final AbstractProject<?, ?>
...
+         * Indicates whether or not this scanning tool has a default pattern. If it
+         * does, it means it can never scan the console, but also means that we don't
+         * require a user-specified pattern as we have a usable default.
+         * 
+         * @return true if {@link #getPattern()} returns a non-empty string.
+         */
+        public boolean hasDefaultPattern() {
+            // Maintenance note: We must use the same "is this empty/blank" logic as used in the runtime code.
+            final String defaultPattern = getPattern();
+            return !StringUtils.isBlank(defaultPattern);
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/analysis/core/model/ReportScanningTool.java:317]:
```suggestion
            return StringUtils.isNotBlank(defaultPattern);
```

File: plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/config.jelly [hunk 21]
```diff
@@ -4,15 +4,24 @@
   <i:tool-help/>
 
   <j:choose>
-    <j:when test="${size(descriptor.pattern) > 0}">
+    <j:when test="${descriptor.hasDefaultPattern()}">
       <f:entry title="${%title.pattern}" field="pattern"
-               description="${%description.overridePattern(descriptor.pattern)}">
+               description="${%description.patternOrDefault('http://ant.apache.org/manual/Types/fileset.html', descriptor.pattern)}"
+               help="${descriptor.getHelpFile('patternOrDefault')}">
```
Comment (uhafner) [inline plugin/src/main/resources/io/jenkins/plugins/analysis/core/model/ReportScanningTool/config.jelly:10]: Good idea! I did not know that this can be changed...

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly (hunk 13 above)
Comment (uhafner) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/config.jelly:16]: Yes, this has been replaced in Jenkins core recently and I forgot to change the documentation as well.

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html (hunk 14 above)
Comment (uhafner) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:13]: @daniel-beck @Wadeck Is this ok for you?

**Review by uhafner — APPROVED** [review]
Thanks for this high quality PR!

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html (hunk 19 above)
Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:6]: 🤷 It was deliberate, yes ... and I know that the resulting message renders ok, but I don't know whether or not it would be better to have that middle sentence within its own `<p>` ... `</p>` section too.
i.e. deliberate ... but potentially misguided 😉

File: plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html (hunk 14 above)
Comment (daniel-beck) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:13]: As I wrote elsewhere, I think we need to revisit the security considerations for Overall/Administer users after we retired the RunScripts permission around Jenkins 2.222.

That said, I recommend you consider whether you want to allow users with Overall/Manage permission (JEP-223) to configure these parsers at some point in the near future.

FWIW the real long term solution for all of this seems to be Script Security integration anyway. Then none of this is a concern -- an admin will not be impacted (their scripts will be pre-approved), Overall/Manage users' parsers need to be approved by admins first or sandboxes, and the scripts can run wherever.

Comment (pjdarton) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:13]: Yes  I view this PR as only being useful until this plug-in has sandboxed scripts.
(and if I hadn't been in a big hurry to "just get something working now!" then I would've looked into that instead).
Once scripts are sandboxed then this configuration option is irrelevant and can be hard coded as "on".

Comment (uhafner) [reply plugin/src/main/resources/io/jenkins/plugins/analysis/warnings/groovy/ParserConfiguration/help-consoleLogScanningPermitted.html:13]: I implemented the script security integration already some years ago. I removed that part because then the parsers did not work on the agent anymore. Script security is only available on the controller side.
